### ISSUE TITLE
Hovering over dropdown arrow in detail sidebar rapidly flashes mouse cursor

### ISSUE THREAD

**Issue by rohanc** [jakob/Postico#405]
I'm seeing some unusual behaviour in the Detail Sidebar when I'm preparing to edit the contents of some table fields. When I move my mouse cursor to certain locations (without clicking) I see the mouse cursor rapidly switches back and forth between the arrow and the "I bar" (text entry) cursor. It seems to the user that the cursor is being captured, as it takes far more hand movement than usual to move the cursor a small distance vertically. It has the subjective feel of decreased mouse acceleration, but only in the detail sidebar area.

It happens with text fields, when my mouse cursor is positioned above the down arrow button which opens the "NULL / DEFAULT / Empty String" menu. It also happens with integer[] fields because they have a similar type of menu. It doesn't happen if my mouse cursor hovers above the up+down arrows on the right of a boolean field.

The behaviour only happens when the Postico app window has focus. If I give focus to another window, by clicking, and the go back and hover over the down arrow icon in the Postico detail sidebar, the mouse behaviour is normal. i.e. the background colour of the arrow icon changes to grey on hover and back to white again when I move away. No mouse capture cursoring occurs.

I've not noticed this behaviour in any other apps on my Macbook Pro. The flashing between two different mouse cursors is most pronounced on my external 4K monitor, but the mouse cursor weirdness still occurs on the inbuilt Retina screen. Trackpad and mouse both exhibit this behaviour and it still happens when the mouse is unplugged. OS is Sierra 10.12.5, but this behaviour has been with me for a while. I use Postico v.1.1.2.

**Comment by jakob** [comment]
I just tried it as well, and I also see some occasional flickering on my Macbook Pro (both internal and on external display). It is probably caused by placing an NSButton on top of an NSTextView, but I am not sure. The effect is barely noticeable on my MacBook Pro (Retina, 13-inch, Early 2015) -- what exact model do you have? Does it have a discrete GPU? Are you using built-in trackpad, a magic mouse, or something different?

**Comment by rohanc** [comment]
MacBook Pro (Retina, 13-inch, Early 2015). No discrete GPU, just the integrated Iris Graphics 6100. Seems to happen in all Retina scaling modes.

**Comment by jakob** [comment]
I have fixed this in build 1771
https://eggerapps.at/postico/builds/

Thanks for reporting this issue!

**Comment by rohanc** [comment]
Thanks Jakob, after some further testing I can recommend to others in my situation to **close their laptop lid** when using an external display. Here's the details:

1. Retina Macbook lid open, 4K external monitor plugged in using displayport cable, run Postico v1.1.2: Terrible mouse cursor issues in Detail Sidebar, especially when moving mouse cursor vertically, if it passes over those arrow icons. The mouse cursor is 'captured' at times and hard to move away from certain spots.

2. Retina Macbook lid open, 4K external monitor plugged in using displayport cable, run Postico build 1771: Capturing of mouse cursor no longer occurs, much better mouse cursor performance, but still noticeable slowdown in mouse cursor movement verses amount of hand movement when moving vertically over those fields.

3. Retina Macbook lid closed, 4K external monitor plugged in using displayport cable, run Postico v1.1.2: mouse cursor movement seems more or less normal.

In other words, your fix in build 1771 is much appreciated, but there would appear to be some horrible display driver bugs in the OS. It's fair enough that the weak GPU in my macbook is slow at rendering the screen when it has to support multiple hi res displays, but it's not OK for the OS to throw away mouse movements, such that the mouse cursor moves half the normal distance (and never catches up) when the GPU is under load.

It's possible that there might be some mouse/trackpad driver software interfering on my system -- I've tried to remove them all (nothing obvious shown by kextstat), but I'd need a full reinstall to be 100% sure.

**Comment by jakob** [comment]
OK, so maybe there are two different issues?

**Issue 1: Mouse Cursor Flickering**
This was caused by a bug in my code, and I fixed it in build 1771. This would happen with any pointing device (mouse or trackpad).

**Issue 2: Odd mouse movement**
I can not reproduce this. Does this happen with both your trackpad and your mouse?

Theoretically my code should not affect mouse cursor movement, since mouse movement / drawing the mouse cursor happens at a much lower level than AppKit. But I do have a far fetched possible explanation: the text fields track mouse movements, so every time the mouse pointer moves a pixel, CPU usage will spike, which can cause voltage fluctuations. These voltage fluctuations can have odd side effects -- for example I had a power supply that made a high pitched noise any time I scrolled a web page.

If the odd movement happens only with your mouse, it is possible that your mouse is sensitive to these voltage fluctuations. Trying a different mouse, or plugging the mouse into a powered USB hub, could check if this is the case.

I do concede that this explanation is a bit far fetched, but far fetched things do tend to happen with electronics :)

**Comment by rohanc** [comment]
With my laptop lid open and a 4K screen connected, I definitely get odd mouse cursor movement when hovering of the Detail sidebar, even when using my trackpad and with my mouse unplugged. It makes no difference whether Postico is placed on the desktop on the laptop screen or on the external screen.

If I unplug the 4K screen, normal mouse movement returns. Or I just close the laptop lid so that the Macbook is driving just one screen, that also fixes it.

I would conjecture that your code can influence how much work the CPU has to do as the mouse moves across a region, but unless you have code that relocates the mouse cursor, I agree that the problem probably lies at a lower level. I suspect OS drivers are at fault, not hardware.
